File: 409-longest-palindrome/longest-palindrome.js
var longestPalindrome = function(s) {
    const count = new Map();

    for (const ch of s) {
        count.set(ch, (count.get(ch) || 0) + 1);
    }

    let length = 0;
    let hasOdd = false;

    for (const freq of count.values()) {
        length += Math.floor(freq / 2) * 2;

        if (freq % 2 === 1) {
            hasOdd = true;
        }
    }

    if (hasOdd) {
        length++;
    }

    return length;
};

File: 409-longest-palindrome/longest-palindrome.cpp
class Solution {
public:
    int longestPalindrome(string s) {
        unordered_map<char, int> count;

        for (char c : s) {
            count[c]++;
        }

        int result = 0;
        bool odd = false;

        for (auto& [ch, freq] : count) {
            result += (freq / 2) * 2;

            if (freq % 2 == 1) {
                odd = true;
            }
        }

        if (odd) {
            result++;
        }

        return result;
    }
};